### PR TITLE
Training stage

### PR DESCRIPTION
**Pull request by yangtcai**
related to #12 
I will update in this pr
- [x] split the datasets for train
- [x] add TensorBoard

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -237,19 +238,49 @@ def test_criterion():
...
+    transformer = build_transformer()
+
+    model = DETR(
+        transformer,
+        num_classes=num_classes,
+        num_queries=num_queries,
+    )
+    matcher = build_matcher()
+    losses = ["classes", "coordinates"]
+    weight_dict = {"loss_ce": 1, "loss_ssegments": 1, "loss_IOU": 1}
```
Comment (yangtcai) [inline model.py:265]: It will be hyperparamaters

File: transformer.py [hunk 2]
```diff
@@ -377,9 +377,13 @@ def _get_activation_fn(activation):
     raise RuntimeError(f"activation should be relu/gelu, not {activation}.")
 
 
+def build_transformer():
+    return Transformer(d_model=6, nhead=6)
+
+
 if __name__ == "__main__":
     n, s, e = 10, 100, 5
-    transformer = Transformer(d_model=5, nhead=5)
```
Comment (williamstark01) [inline transformer.py:382]: This looks necessary for the test?

File: train.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    model, criterion = build_model()
+    model.to(device)
+    criterion.to(device)
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=0.001)
+
+    dataloader = build_dataloader()
+
+    for epoch in range(2):
```
Comment (williamstark01) [inline train.py:67]: Probably meant:
```
    for epoch in range(1):
```

**Review by williamstark01 — COMMENTED** [review]
Good job writing your own training loop. Some logging will help with tracking the training process, and as you mention, TensorBoard can also help with that.

加油！

File: train.py (hunk 3 above)
Comment (yangtcai) [reply train.py]: I have wrap it into args :D

**Review by williamstark01 — APPROVED** [review]
We discussed about segments (subsequences) which don't contain repeats before and I'm trying to figure out the best way to skip over them during training. Maybe run forward_strand for all indexes in the RepeatSequenceDataset initialization stage and saving a list of those with repeats?

Did you track loss during training tests to see whether it's decreasing?

File: dataloader.py [hunk 4]
```diff
@@ -111,9 +117,34 @@ def build_dataset():
...
+    )
+    validation_loader = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=args.batch_size,
+        sampler=valid_sampler,
+        collate_fn=dataset.collate_fn,
+    )
+    return DataLoader(
+        dataset, batch_size=args.batch_size, shuffle=True, collate_fn=dataset.collate_fn
+    )
```
[the remark this answers is not in the document]
Comment (yangtcai) [reply dataloader.py:147]: > 
Sounds great! I will try it later.
I have tracked the loss on my own computer, the loss looks descending. However, it's too slow, so I have run litter parts.

**Comment by yangtcai** [comment]
Hi, @williamstark01, the advice sounds great! I will try it later.
I have tracked the loss on my own computer, the loss looks descending. However, it's too slow, so I have run litter parts.
